File: compose_version/app/src/main/java/com/example/news_app/domain/usecase/GetTopHeadlinesUseCase.kt
package com.example.news_app.domain.usecase

import com.example.news_app.domain.entities.ArticlesPage
import com.example.news_app.domain.repository.NewsRepository

class GetTopHeadlinesUseCase(private val repository: NewsRepository) {
    suspend fun execute(
        country: String = "us",
        category: String? = null,
        q: String? = null,
        page: Int = 1,
        pageSize: Int = 20,
    ): ArticlesPage {
        return repository.getTopHeadlines(
            country = country,
            category = category,
            q = q,
            page = page,
            pageSize = pageSize,
        )
    }
}

File: swiftui_version/News/News/Features/News/Domain/Usecases/GetTopHeadlinesUseCase.swift
import Foundation

/// Use case for fetching top headline articles
class GetTopHeadlinesUseCase {
    private let repository: NewsRepository
    
    init(repository: NewsRepository) {
        self.repository = repository
    }
    
    func call(
        country: String = "us",
        category: String? = nil,
        query: String? = nil,
        page: Int = 1,
        pageSize: Int = 20
    ) async throws -> ArticlesPage {
        AppLogger.d("GetTopHeadlinesUseCase: calling with country=\(country), category=\(category ?? "nil"), query=\(query ?? "nil"), page=\(page)")
        
        return try await repository.getTopHeadlines(
            country: country,
            category: category,
            query: query,
            page: page,
            pageSize: pageSize
        )
    }
}
